### PR TITLE
Add additional call functions for core other ops

### PR DESCRIPTION
**Pull request by cak**
Fixed calls to core other operations by adding two additional call functions: 
- `call_other` for text responses
- `call_other_json` for JSON responses

Added relevant `scan_report.rs` to /examples.

### CODE REVIEW

File: src/core.rs [hunk 1]
```diff
@@ -928,55 +928,55 @@ pub fn set_option_use_proxy_chain_auth(
     )
 }
 
-pub fn proxy_pac(service: &ZapService) -> Result<Value, ZapApiError> {
+pub fn proxy_pac(service: &ZapService) -> Result<String, ZapApiError> {
```
Comment (psiinon) [inline src/core.rs]: This file (and most of the others) are actually generated by this java class: https://github.com/zaproxy/zaproxy/blob/develop/zap/src/main/java/org/zaproxy/zap/extension/api/RustAPIGenerator.java

So we'll need to change that and then regenerate all of the relevant files.
Are you ok having a go at those changes? If not then I can look at them :)

File: examples/scan_report.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::fs;
+use std::{thread, time};
+use zap_api::ZapApiError;
+use zap_api::ZapService;
+
+fn main() -> Result<(), ZapApiError> {
```
Comment (psiinon) [inline examples/scan_report.rs]: Looks good :) Do we still need https://github.com/zaproxy/zap-api-rust/blob/master/examples/main.rs ? This looks like its a superset of that.

File: src/core.rs (hunk 1 above)
Comment (cak) [reply src/core.rs]: I was wondering about the "This file was automatically generated." - the generator is awesome! 

Unfortunately, I am not familiar with Java, however if you wanted to keep it simple with one `fn call...`  for [return](https://github.com/zaproxy/zaproxy/blob/49ceadcd92e26d9d91b9470877f9c5471bf22696/zap/src/main/java/org/zaproxy/zap/extension/api/RustAPIGenerator.java#L150) and [call](https://github.com/zaproxy/zaproxy/blob/49ceadcd92e26d9d91b9470877f9c5471bf22696/zap/src/main/java/org/zaproxy/zap/extension/api/RustAPIGenerator.java#L179) there is an example option: 

Updated call function in `lib.rs` and ZapResponse enum wrapper:

``` Rust
pub enum ZapResponse {
    Text(String),
    Value(Value),
}

impl ZapResponse {
    pub fn text(self) -> Option<String> {
        match self {
            ZapResponse::Text(text) => Some(text),
            _ => None,
        }
    }

    pub fn value(self) -> Option<Value> {
        match self {
            ZapResponse::Value(value) => Some(value),
            _ => None,
        }
    }
}

pub fn call_option(
    service: &ZapService,
    op: &str, // Adding "OTHER" or "JSON"
    component: &str,
    calltype: &str,
    method: &str,
    json: bool, // Return Value if true, String if false
    _params: HashMap<String, String>,
) -> Result<ZapResponse, ZapApiError> {
    let mut url = [&service.url, op, component, calltype, method, ""].join("/");
    if _params.keys().len() > 0 {
        url.push_str("?");
        for (key, value) in _params {
            url.push_str(&key);
            url.push_str("=");
            url.push_str(&value);
            url.push_str("&");
        }
    }

    let client = reqwest::Client::new();
    let text = client
        .get(&url)
        .header("X-ZAP-API-Key", &*service.api_key)
        .send()?
        .text()?;
    if json {
        let value = ZapResponse::Value(serde_json::from_str(&text)?);
        Ok(value)
    } else {
        let string = ZapResponse::Text(text); 
        Ok(string)
    }
}
```

Making the changes in the other files (core.rs examples) :
```Rust
pub fn htmlreport(service: &ZapService) -> Result<ZapResponse, ZapApiError> {
    let params = HashMap::new();
    //super::call_other(service, "core", "other", "htmlreport", params)
    super::call_option(service, "OTHER", "core", "other", "htmlreport", false, params)
}

pub fn jsonreport(service: &ZapService) -> Result<ZapResponse, ZapApiError> {
    let params = HashMap::new();
    //super::call_other_json(service, "core", "other", "jsonreport", params)
    super::call(service, "OTHER", "core", "other", "jsonreport", true, params)
}
```

The API user would then use the following (downside is they would have to call the `.text()` or `.value()` method and unwrap:

``` Rust
fn save_report_html(service: &ZapService) -> Result<(), ZapApiError> {
    println!("Saving the HTML report");
    let res = zap_api::core::htmlreport(&service);
    match res {
        Ok(v) => {
            let report_data = v.text().unwrap();
            fs::write("zap_report.html", report_data).expect("Unable to save ZAP report");
        }
        Err(e) => return Err(e),
    }

    Ok(())
}
```

File: examples/scan_report.rs (hunk 2 above)
Comment (cak) [reply examples/scan_report.rs]: Correct, this example includes and expands on `main.rs`'s functionality. :)

File: src/core.rs (hunk 1 above)
Comment (cak) [reply src/core.rs]: Reverted changes from `core.rs` which the ZAP Core Team will regenerate.

File: examples/scan_report.rs (hunk 2 above)
Comment (cak) [reply examples/scan_report.rs]: Removed `main.rs`, replaced by `scan_report.rs`.

**Comment by cak** [comment]
Thank you to @psiinon (and ZAP Core Team) for taking over the regeneration of the `core.rs` file.  🚀

The following changes are required:

- proxy.pac (proxy_pac) -> call_other
- rootcert -> call_other
- setproxy -> call_other
- xmlreport -> call_other
- htmlreport -> call_other
- jsonreport -> call_other_json
- mdreport -> call_other
- messageHar (message_har) -> call_other_json
- messagesHar (messages_har) -> call_other_json
- messagesHarById (messages_har_by_id) -> call_other_json
- sendHarRequest (send_har_request) -> call_other_json

Additionally, the  function return signatures using `call_other` will need to be modified to `Result<String, ZapApiError>` (`call_other_json` can remain unchanged with `Result<Value, ZapApiError>`) and 

Examples:

``` Rust
pub fn mdreport(service: &ZapService) -> Result<String, ZapApiError> {
    let params = HashMap::new();
    super::call_other(service, "core", "other", "mdreport", params)
}

pub fn message_har(service: &ZapService, id: String) -> Result<Value, ZapApiError> {
    let mut params = HashMap::new();
    params.insert("id".to_string(), id);
    super::call_other_json(service, "core", "other", "messageHar", params)
}
```

**Comment by cak** [comment]
Closing this PR for now, will create a new one soon to upgrade [reqwest](https://github.com/seanmonstar/reqwest) and update for async/await.

**Pull request closed without merging**